### PR TITLE
Add a trace and missing arguments to riem_measures, per issue 23

### PR DESCRIPTION
**Pull request by mattsq**
Hey - giving back to a very useful project - thought I'd have a crack at #23 - let me know your thoughts.

### CODE REVIEW

File: R/measures.R [hunk 1]
```diff
@@ -47,16 +49,26 @@
...
+  base_link <- "https://mesonet.agron.iastate.edu/cgi-bin/request/asos.py/"
 
+  if(!is.na(trace) & !(trace %in% list("T", "", .0001))) {
+    stop(call. = FALSE,
+         "trace must be one of NA, 'T', '' or 0.0001.")  # nolint
+  }
 
+  if(!is.na(missing) & !(missing %in% list("M", ""))) {
+    stop(call. = FALSE,
+         "missing must be one of NA, 'M', ''")  # nolint
```
Comment (maelle) [inline R/measures.R]: I don't understand, I had read "M, null, or empty"?

Comment (maelle) [reply R/measures.R]: Oh I guess that's what happens to the JSON? I suppose a test would help me see what's expected.

**Review by maelle — CHANGES_REQUESTED** [review]
Awesome, thank you!! Could you please

* add a test / a few lines in the existing test (happy to provide more pointers if needed)
* update NEWS.md
* add yourself as "ctb" in DESCRIPTION?

Also note that I'll be on vacation during the next 3 weeks so might take a while to respond.

File: R/measures.R (hunk 1 above)
Comment (mattsq) [reply R/measures.R]: OK, so - turns out I misread the original issue #22 , but I caught up! It now correctly passes those arguments (in their intended form) along to the GET request, and then handles the output. I've also added tests, and updated the DESCRIPTION and the NEWS file. Happy to add more tests if you think we should cover it more exhaustively.

File: NEWS.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
+# riem 0.1.2
+
+* Handle the new 'missing' and 'trace' parameters made available by IEM
+  * Defaults are as current behaviour, but you can now return other values for missing data and trace preciptation
```
Comment (maelle) [inline NEWS.md:4]:
```suggestion
  * Defaults are as current behaviour, but you can now return other values for missing data and trace precipitation
```

File: R/measures.R [hunk 3]
```diff
@@ -47,16 +49,26 @@
 #'
 #' @examples
 #' \dontrun{
-#' riem_measures(station = "VOHY", date_start = "2000-01-01", date_end = "2016-04-22")
+#' riem_measures(station = "VOHY", date_start = "2000-01-01", date_end = "2016-04-22", trace = 1e-4, missing = "empty")
```
Comment (maelle) [inline R/measures.R:52]: I think it'd make sense to choose either 1e-4 or .0001 as notation in the whole file.

File: R/measures.R [hunk 4]
```diff
@@ -47,16 +49,26 @@
...
-                          date_end = as.character(Sys.Date())){
+                          date_end = as.character(Sys.Date()),
+                          trace = .0001,
+                          missing = "empty"){
 
-  base_link <- "https://mesonet.agron.iastate.edu/cgi-bin/request/asos.py/"
 
+  base_link <- "https://mesonet.agron.iastate.edu/cgi-bin/request/asos.py/"
 
+  if(!(trace %in% list("T", "null", "empty", .0001))) {
```
Comment (maelle) [inline R/measures.R:63]: Oh, I see, it is `list` because of the different types?

File: R/measures.R [hunk 5]
```diff
@@ -47,16 +49,26 @@
...
+    stop(call. = FALSE,
+         "trace must be one of 'T', 'null', 'empty' or 0.0001.")  # nolint
+  }
 
+  if(!(missing %in% list("M", "empty","null"))) {
```
Comment (maelle) [inline R/measures.R:68]:
```suggestion
  if(!(missing %in% c("M", "empty","null"))) {
```

That one can be `c()`?

File: R/measures.R [hunk 6]
```diff
@@ -87,13 +99,15 @@ riem_measures <- function(station = "VOHY",
...
+                                 latlon = "yes",
+                                 missing = missing,
+                                 trace = trace))
   content <- httr::content(page)
 
   col.names <- t(suppressWarnings(read.table(text = content,
                                            skip = 5,
                                            nrows = 1,
-                                           na.strings = c("", "NA", "M"),
+                                           na.strings = c("NA"),
```
Comment (maelle) [inline R/measures.R:110]: Could you remind me why you had to change this line?

File: tests/testthat/test-measures.R [hunk 7]
```diff
@@ -51,3 +51,16 @@ test_that("riem_measures checks dates",{
                "date_end has to be bigger than date_start")
 })
 
+test_that("riem_measures returns correct values when alternate arguments to missing are passed",{
+  skip_on_cran()
+  output <- riem_measures(station = "VOHY", date_start = "2014-01-01",
+                          date_end = "2016-04-22", missing = "M")
```
Comment (maelle) [inline tests/testthat/test-measures.R:57]: So when you choose "M", R will consider the whole column to be character, right? So why would one do that, actually?

**Review by maelle — CHANGES_REQUESTED** [review]
Thank you! I made a few comments, the last one being the most important (I should have thought of this when opening the issue): why would one choose the alternative values, and when one chooses them, how problematic is it to make the whole column character? Do you have any use case?

**Comment by maelle** [comment]
I'll try to come back to this in the new year. Note that there's been a rather big change in the default branch, the package now uses the new web services.

**Comment by maelle** [comment]
Closing because there has been much refactoring on the main branch.

**Pull request closed without merging**